### PR TITLE
Cassandra product tests

### PR DESCRIPTION
**Pull request by brian-rickman**
@cawallin @maciejgrzybek

### CODE REVIEW

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypes.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+public class DataTypes
+        extends ProductTest
+        implements RequirementsProvider
+{
+    private static final String ALL_TYPES_TABLE_NAME = "all_types";
+
+    private static final CassandraTableDefinition ALL_TYPES_TABLE_DEFINITION;
+
+    // Leaving out "bl blob" because I don't know how to represent it in Java
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypes.java]: It's most likely just a `byte[]`.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypes.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                                "abc", new BigInteger("567")),
+                        ImmutableList.of("def", 123L, false, new BigDecimal(456.789), 1234.5678,
+                                90123.456f, ImmutableSet.of(1, 2, 3), Inet4Address.getByName("127.0.0.1"), 1, ImmutableList.of(4, 5, 6), ImmutableMap.of("a", 1, "b", 2),
+                                ImmutableSet.of(4, 5, 6), "this is a text value", Timestamp.valueOf("2011-02-03 04:05:00"),
+                                UUID.fromString("d2177dd0-eaa2-11de-a572-001b779c76e3"), UUID.fromString("01234567-0123-0123-0123-0123456789ab"),
+                                "abc", new BigInteger("567"))
+                ).iterator();
+            }
+            catch (UnknownHostException e) {
+                e.printStackTrace();
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypes.java]: `Throwables.propagate` instead maybe?
We use it in I believe tempto-examples in the same context.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypes.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    static {
+        RelationalDataSource dataSource = () -> {
+            try {
+                return ImmutableList.<List<Object>>of(
+                        ImmutableList.of("abc", 123L, false, new BigDecimal(456.789), 1234.5678,
+                                90123.456f, ImmutableSet.of(1, 2, 3), Inet4Address.getByName("127.0.0.1"), 1, ImmutableList.of(4, 5, 6), ImmutableMap.of("a", 1, "b", 2),
+                                ImmutableSet.of(4, 5, 6), "this is a text value", Timestamp.valueOf("2011-02-03 04:05:00"),
+                                UUID.fromString("d2177dd0-eaa2-11de-a572-001b779c76e3"), UUID.fromString("01234567-0123-0123-0123-0123456789ab"),
+                                "abc", new BigInteger("567")),
+                        ImmutableList.of("def", 123L, false, new BigDecimal(456.789), 1234.5678,
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypes.java]: Why not test edge cases as well, e.g. `Double.Infinity` etc.?

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
+/*
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java]: What's special about `Bcookie` table? Where this name comes from?

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/Select.java [hunk 5]
```diff
@@ -56,4 +61,95 @@ public void testSelectNation()
...
+                KEY_SPACE,
+                CASSANDRA_NATION.getName());
+        QueryResult queryResult = onPresto()
+                .executeQuery(sql);
+
+        assertThat(queryResult).containsOnly(row("UNITED STATES"));
+    }
+
+    @Test(groups = CASSANDRA)
+    public void testSelectWithFilterOnNonPartitioningKey()
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/Select.java]: `withNonEqualityFilter` to be consistent with previous tests.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/JoinWithTpch.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        implements RequirementsProvider
+{
+    @Override
+    public Requirement getRequirements(Configuration configuration)
+    {
+        return immutableTable(CASSANDRA_NATION);
+    }
+
+    @Test(groups = CASSANDRA)
+    public void testNationJoinRegion()
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/JoinWithTpch.java]: How about self-join (additional test, `testNationJoinRegion` is valid too though!)? This would ensure we can run two concurrent connections from the same session to the same remote Cassandra resources.

**Review by maciejgrzybek — CHANGES_REQUESTED** [review]
I don't see any negative tests. I.e. when query is expected to fail. Are there any cases like that?
Corner cases of data types are also not covered.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import static com.facebook.presto.tests.cassandra.TestConstants.CONNECTOR_NAME;
+import static com.facebook.presto.tests.cassandra.TestConstants.KEY_SPACE;
+import static java.sql.JDBCType.TIMESTAMP;
+import static java.sql.JDBCType.VARCHAR;
+
+public class BcookieTableDefinition
+{
+    private BcookieTableDefinition() {}
+
+    // Since the two tables are identical except in name, we'll just define bcookie here.
```
Comment (cawallin) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java]: Remove this comment, it's not necessary.

**Review by cawallin — COMMENTED** [review]
One small comment, I agree with what Maciek said too

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java (hunk 7 above)
Comment (brian-rickman) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java]: done

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java (hunk 4 above)
Comment (brian-rickman) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java]: It represents the schema our customer in Japan will use, but I was told not to mention their name.  I can give the table a more generic name.  The only thing that makes it special for testing is the use of a two column primary key.

Comment (maciejgrzybek) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java]: Please use generic name then, please.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/Select.java (hunk 5 above)
Comment (brian-rickman) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/Select.java]: ok

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java (hunk 4 above)
Comment (brian-rickman) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/BcookieTableDefinition.java]: Renamed to MultiColumnKeyTableDefinition

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/JoinWithTpch.java (hunk 6 above)
Comment (brian-rickman) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/JoinWithTpch.java]: Added self-join test

**Comment by brian-rickman** [comment]
Added a new commit that addresses most of the comments.  Still need to add negative tests and edge cases for datatypes.  Also need to fix blob test.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/MultiColumnKeyTableDefinition.java [hunk 8]
```diff
@@ -26,35 +26,31 @@
...
+    public static final ImmutableList<JDBCType> MULTI_COLUMN_KEY_TYPES = ImmutableList.of(VARCHAR, VARCHAR, VARCHAR, TIMESTAMP);
+    private static final String MULTI_COLUMN_KEY_DDL =
             "CREATE TABLE %NAME% (" +
                     "user_id text, " +
                     "key text, " +
                     "updated_at timestamp, " +
                     "value text, " +
                     "PRIMARY KEY (user_id, key));";
-    private static final String BCOOKIE_TABLE_NAME = "bcookie";
+    private static final String MULTI_COLUMN_KEY_TABLE_NAME = "bcookie";
```
Comment (cawallin) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/MultiColumnKeyTableDefinition.java]: rename the string "bcookie" too

**Review by cawallin — COMMENTED** [review]
updates look good, I'll look out for updates to add negative tests and edge cases
For the blob test, can you just use a byte[]? or new String("0123).toBytes()?

**Comment by brian-rickman** [comment]
@cawallin "For the blob test, can you just use a byte[]? or new String("0123).toBytes()?"
I tried both of those.  The only thing the Cassandra driver accepts is their Bytes class, but of course, that is not what we get back from Presto.

**Comment by cawallin** [comment]
Ahhh yeah. I wonder if you have to do new String("0123").toBytes() for what you get out of Presto.

**Comment by maciejgrzybek** [comment]
@brian-rickman doesn't it help http://stackoverflow.com/a/31818219? I know it's opposite (Cassandra's Byte -> String, instead of String -> Cassandra) but I believe the opposite should be possible given this example.

**Comment by brian-rickman** [comment]
@maciejgrzybek I looked at that page earlier, but I'm not sure how it helps.  I don't think I have direct access to the Cassandra API here as Tempto is doing the insert.  On the return side, we are retrieving the data through Presto.

Tempto not displaying the actual values in the error message makes this so much harder than it should be.  I already filed a ticket for that.

**Comment by maciejgrzybek** [comment]
@brian-rickman data provided in `RelationalDataSource` can be given in native Cassandra types. It does not have to be JDBC.
It's directly passed to Cassandra API.
So you should be able to do something like:
```java
 RelationalDataSource dataSource = () -> {
  return ImmutableList.<List<Object>>of(ByteBuffer.wrap("foo".getBytes()).iterator();
}
```
or something similar.

**Comment by brian-rickman** [comment]
@cawallin @maciejgrzybek I added negative tests.  

I also got the blob test working.  I had read from the wrong column of an ASCII table when I created the test the first time :(.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypesTableDefinition.java [hunk 9]
```diff
@@ -50,11 +50,17 @@ private DataTypesTableDefinition() {}
...
             try {
                 return ImmutableList.<List<Object>>of(
-                        ImmutableList.of("abc", 123L, Bytes.fromHexString("0x48495051"), false, new BigDecimal(456.789), 1234.5678,
+                        ImmutableList.of("abc", 123L, Bytes.fromHexString("0x30313233"), false, new BigDecimal(456.789), 1234.5678,
                                 90123.456f, ImmutableSet.of(1, 2, 3), Inet4Address.getByName("127.0.0.1"), 1, ImmutableList.of(4, 5, 6), ImmutableMap.of("a", 1, "b", 2),
                                 ImmutableSet.of(4, 5, 6), "this is a text value", Timestamp.valueOf("2011-02-03 04:05:00"),
                                 UUID.fromString("d2177dd0-eaa2-11de-a572-001b779c76e3"), UUID.fromString("01234567-0123-0123-0123-0123456789ab"),
                                 "abc", new BigInteger("567")),
+                        ImmutableList.of("the quick brown fox jumped over the lazy dog", 9223372036854775807L, Bytes.fromHexString("0x3031323334"),
+                                false, BigDecimal.valueOf(Double.MAX_VALUE), Double.MAX_VALUE,
```
Comment (cawallin) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypesTableDefinition.java]: A couple of things:
* Try the highest possible Presto decimal (precision 38, all 9's). It wouldn't be a bad idea to try both long and short decimal from Presto.
* non-ascii characters for varchar and text (and other text-ish fields)
* Long.MAX_VALUE instead of the value
* true instead of false (just to test both)
* Try Bytes.fromHexString("0x00") to see if null is handled properly

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/NegativeTests.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    @Test(groups = CASSANDRA)
+    public void testInvalidTable()
+            throws SQLException
+    {
+        String tableName = format("%s.%s.%s", CONNECTOR_NAME, KEY_SPACE, "bogus");
+        String sql = format("SELECT * FROM %s", tableName);
+        try {
+            onPresto().executeQuery(sql);
+        }
+        catch (QueryExecutionException e) {
```
Comment (cawallin) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/NegativeTests.java]: You can use the following helper method:
assertThat(() -> onPresto.executeQuery(sql)).failsWithMessage(errormessage);

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/NegativeTests.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Test(groups = CASSANDRA)
+    public void testInvalidSchema()
+            throws SQLException
+    {
+        String tableName = format("%s.%s.%s", CONNECTOR_NAME, "does_not_exist", "bogus");
+        String sql = format("SELECT * FROM %s", tableName);
+        try {
+            onPresto().executeQuery(sql);
```
Comment (cawallin) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/NegativeTests.java]: ditto about assertThat

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/NegativeTests.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        catch (QueryExecutionException e) {
+            assertThat(e.getMessage().equals("Schema does_not_exist does not exist"));
+        }
+    }
+
+    @Test(groups = CASSANDRA)
+    public void testInvalidColumn()
+            throws SQLException
+    {
+        String tableName = format("%s.%s.%s", CONNECTOR_NAME, KEY_SPACE, CASSANDRA_NATION.getName());
```
Comment (cawallin) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/NegativeTests.java:65]: ditto about assertThat

**Review by cawallin — COMMENTED** [review]
A few changes to the datatype tests, but this is on the right track. It'd be good to also add min datatype values.

**Comment by brian-rickman** [comment]
Fixed the compile error so the tests can run.  Still have one failing test `com.facebook.presto.tests.cassandra.Select.testAllDatatypes` , and would appreciate advice on what might be wrong with it.

**Comment by brian-rickman** [comment]
Same test is failing, though now the values Tempto displays for Actual and Expected result appear to be identical.  I'm going to take a break from this to work on testing the ODBC drivers. 

2016-11-28 16:39:47 INFO: [1 of 17] com.facebook.presto.tests.cassandra.Select.t                                                                             estAllDatatypes (Groups: cassandra)


*********** DEBUG ***********
java.lang.String:
java.lang.Long: -9223372036854775808
byte[]: [B@488aa99c
Byte array value: 0 Bytes.fromHexString("0x00").array(): 0
java.lang.Boolean: false
java.lang.Double: 0.0
java.lang.Double: 4.9E-324
java.lang.Double: 1.401298464324817E-45
java.lang.String: [0]
java.lang.String: 0.0.0.0
java.lang.Integer: -2147483648
java.lang.String: [0]
java.lang.String: {"\u0000":-2147483648,"a":0}
java.lang.String: [0]
java.lang.String:
java.sql.Timestamp: 1970-01-01 00:00:00.0
java.lang.String: d2177dd0-eaa2-11de-a572-001b779c76e3
java.lang.String: 01234567-0123-0123-0123-0123456789ab
java.lang.String:
java.lang.String: -9223372036854775808


2016-11-28 16:39:53 INFO: FAILURE     /    took 6.4 seconds
2016-11-28 16:39:53 SEVERE: Failure cause:
java.lang.AssertionError: Could not find rows:
[, -9223372036854775808, [B@6880c591, false, 0.0, 4.9E-324, 1.401298464324817E-45, [0], 0.0.0.0, -2147483648, [0], {"\u0000":-2147483648,"a":0}, [0], , 1970-01-01 00:00:00.0, d2177dd0-eaa2-11de-a572-001b779c76e3, 01234567-0123-0123-0123-0123456789ab, , -9223372036854775808]

actual rows:
[the quick brown fox jumped over the lazy dog, 9223372036854775807, [B@880cd78,                                                                              true, 1.0E38, 1.7976931348623157E308, 3.4028234663852886E38, [4,5,6,7], 255.255.255.255, 2147483647, [4,5,6], {"a":1,"b":2}, [4,5,6], this is a text value, 9999-12-31 23:59:59.0, d2177dd0-eaa2-11de-a572-001b779c76e3, 01234567-0123-0123-0123                                                                             -0123456789ab, abc, 9223372036854775807]
[def, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null]
[, -9223372036854775808, [B@488aa99c, false, 0.0, 4.9E-324, 1.401298464324817E-45, [0], 0.0.0.0, -2147483648, [0], {"\u0000":-2147483648,"a":0}, [0], , 1970-01-01 00:00:00.0, d2177dd0-eaa2-11de-a572-001b779c76e3, 01234567-0123-0123-0123-0123456789ab, , -9223372036854775808]
        at org.assertj.core.api.AbstractAssert.failWithMessage(AbstractAssert.java:114)
        at com.teradata.tempto.assertions.QueryAssert.contains(QueryAssert.java:176)
        at com.teradata.tempto.assertions.QueryAssert.containsOnly(QueryAssert.java:201)
        at com.teradata.tempto.assertions.QueryAssert.containsOnly(QueryAssert.java:213)
        at com.facebook.presto.tests.cassandra.Select.testAllDatatypes(Select.java:207)

**Comment by cawallin** [comment]
@maciejgrzybek do you still want any changes? I merged the commits together after fixing the select test that was failing.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypesTableDefinition.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+                                UUID.fromString("01234567-0123-0123-0123-0123456789ab"),
+                                "\0", BigInteger.valueOf(Long.MIN_VALUE)),
+                        ImmutableList.of("the quick brown fox jumped over the lazy dog", Long.MAX_VALUE,
+                                Bytes.fromHexString("0x3031323334"), true,
+                                new BigDecimal(new Double("99999999999999999999999999999999999999")),
+                                Double.MAX_VALUE, Float.MAX_VALUE, ImmutableSet.of(4, 5, 6, 7),
+                                Inet4Address.getByName("255.255.255.255"), Integer.MAX_VALUE,
+                                ImmutableList.of(4, 5, 6), ImmutableMap.of("a", 1, "b", 2), ImmutableSet.of(4, 5, 6),
+                                "this is a text value", Timestamp.valueOf("9999-12-31 23:59:59"),
+                                UUID.fromString("d2177dd0-eaa2-11de-a572-001b779c76e3"),
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypesTableDefinition.java:68]: How about extracting these to constants?

Or maybe a helper method like:
```java
public static Object getColumnValue(CassandraTableDefinition tableDefinition, int colNum, int rowNum) {
List<List<Object>> columns = ImmutableList.copyOf(tableDefinition.getDataSource().getDataRows());
return columns.get(rowNum).get(colNum);
}
```
Then in test you can refer to expected results as, e.g.:
```java
getColumnValue(CASSANDRA_ALL_TYPES, 0, 3);
```
to retrieve expected value from 1st row and 4th column.

I won't insist though.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/Select.java [hunk 14]
```diff
@@ -56,4 +74,124 @@ public void testSelectNation()
...
+
+        assertThat(query)
+                .hasColumns(LONGNVARCHAR, BIGINT, LONGVARBINARY, BOOLEAN, DOUBLE, DOUBLE, DOUBLE, LONGNVARCHAR, LONGNVARCHAR,
+                        INTEGER, LONGNVARCHAR, LONGNVARCHAR, LONGNVARCHAR, LONGNVARCHAR, TIMESTAMP, LONGNVARCHAR, LONGNVARCHAR,
+                        LONGNVARCHAR, LONGNVARCHAR)
+                .containsOnly(
+                        row("\0", Long.MIN_VALUE, Bytes.fromHexString("0x00").array(), false, 0f, Double.MIN_VALUE,
+                                (double) Float.MIN_VALUE, "[0]", "0.0.0.0", Integer.MIN_VALUE, "[0]", "{\"\\u0000\":-2147483648,\"a\":0}",
+                                "[0]", "\0", Timestamp.valueOf("1970-01-01 00:00:00.0"),
+                                "d2177dd0-eaa2-11de-a572-001b779c76e3", "01234567-0123-0123-0123-0123456789ab",
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/Select.java:186]: Have a look on comment about extracting constants above.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/MultiColumnKeyTableDefinition.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    private MultiColumnKeyTableDefinition() {}
+
+    public static final ImmutableList<JDBCType> MULTI_COLUMN_KEY_TYPES = ImmutableList.of(VARCHAR, VARCHAR, VARCHAR, TIMESTAMP);
+    private static final String MULTI_COLUMN_KEY_DDL =
+            "CREATE TABLE %NAME% (" +
+                    "user_id text, " +
+                    "key text, " +
+                    "updated_at timestamp, " +
+                    "value text, " +
+                    "PRIMARY KEY (user_id, key));";
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/MultiColumnKeyTableDefinition.java]: How about adding one more column in key (having partitioning key = { user_id } and compound clustering key, e.g. { key, updated_at })?
This would enable us to test filtering by a prefix of key (which is different path use case than scanning by entire clustering key).
Then we should also test scanning on a column which is part of clustering key but is not a prefix -> Cassandra cannot help with scanning here.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/SelfJoin.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+        implements RequirementsProvider
+{
+    @Override
+    public Requirement getRequirements(Configuration configuration)
+    {
+        return immutableTable(CASSANDRA_NATION);
+    }
+
+    @Test(groups = CASSANDRA)
+    public void testNationJoinNation()
```
Comment (maciejgrzybek) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/SelfJoin.java]: Is it worth to keep it as a separate file?

**Review by maciejgrzybek — CHANGES_REQUESTED** [review]

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypesTableDefinition.java (hunk 13 above)
Comment (cawallin) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypesTableDefinition.java:68]: Similar to Andrii's comment in the integration tests, the types are different so you can't have one source of truth, unfortunately.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/SelfJoin.java (hunk 16 above)
Comment (cawallin) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/SelfJoin.java]: Probably not. I'll change that.

**Comment by cawallin** [comment]
@maciejgrzybek updated

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypes.java (hunk 3 above)
Comment (cawallin) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypes.java]: IMO that's something we should test at a lower level than product tests (e.g. the integration tests should test that).

Comment (maciejgrzybek) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/DataTypes.java]: Given we don't have them, it's better to have it here then don't have it at all.
